hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/upgrade/UpgradeManager.java

```diff
@@ -0,0 +1,204 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.apache.hadoop.hdds.scm.cli.container.upgrade.UpgradeUtils.getContainerDBPath;
+
+/**
+ * This class manage upgrade v2 to v3 container.
+ */
+public class UpgradeManager {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(UpgradeManager.class);
+
+  private final Map<String, DatanodeStoreSchemaThreeImpl>
+      volumeStoreMap = new HashMap<>();
+
+  public void run(OzoneConfiguration configuration) throws IOException {
+    final DatanodeDetails detail =
+        UpgradeUtils.getDatanodeDetails(configuration);
+
+    final MutableVolumeSet dataVolumeSet = UpgradeUtils
+        .getHddsVolumes(configuration, StorageVolume.VolumeType.DATA_VOLUME,
+            detail.getUuidString());
+
+    for (StorageVolume storageVolume : dataVolumeSet.getVolumesList()) {
+      final HddsVolume volume = (HddsVolume) storageVolume;
+      volume.loadDbStore(false);
+    }
+    initVolumeStoreMap(dataVolumeSet, configuration);
+    upgradeAll(dataVolumeSet, configuration);
+  }
+
+  public void initVolumeStoreMap(MutableVolumeSet dataVolumeSet,
+                                 OzoneConfiguration configuration)
+      throws IOException {
+    for (StorageVolume storageVolume : dataVolumeSet.getVolumesList()) {
+      final HddsVolume volume = (HddsVolume) storageVolume;
+      final File containerDBPath = getContainerDBPath(volume);
+      final DatanodeStoreSchemaThreeImpl datanodeStoreSchemaThree =
+          new DatanodeStoreSchemaThreeImpl(configuration,
+              containerDBPath.getAbsolutePath(), false);
+      volumeStoreMap.put(volume.getStorageDir().getAbsolutePath(),
+          datanodeStoreSchemaThree);
+    }
+  }
+
+  public List<Result> upgradeAll(MutableVolumeSet volumeSet,
+                                 OzoneConfiguration configuration) {
+    List<Result> results = new ArrayList<>();
+    final List<StorageVolume> volumesList = volumeSet.getVolumesList();
+    List<CompletableFuture<Result>> volumeFutures =
+        new ArrayList<>(volumesList.size());
+    long startTime = System.currentTimeMillis();
+
+    LOG.info("Start upgrade {} volumes container LayoutVersion",
+        volumesList.size());
+
+    for (StorageVolume volume : volumesList) {
+      final HddsVolume hddsVolume = (HddsVolume) volume;
+      final UpgradeTask upgradeVolume =
+          new UpgradeTask(configuration, hddsVolume,
+              volumeStoreMap.get(hddsVolume.getStorageDir()
+                  .getAbsolutePath()));
+
+      final CompletableFuture<Result> future =
+          upgradeVolume.getUpgradeFutureByVolume();
+      volumeFutures.add(future);
+    }
+
+    for (CompletableFuture<Result> volumeFuture : volumeFutures) {
+      final Result result = volumeFuture.join();
```

Comment:
The join only throws unchecked exception, so once a task failed, the it will exit. 
For example, a DN has 2 volumes, one failed, but another can succeed, I think we should wait it to finish, right? 
So please use get instead of join, and handle the failure cases.